File: treejs_test/neat/genome.ts
import { Gene } from "./gene.ts"
import { GenomeHistory } from "./genome_history.ts"
import { Node } from "./node.ts"
import { Profiler } from "./profiler.ts"

export class Genome {
  private inputs: number
  private outputs: number

  private total_nodes = 0

  private nodes: Node[] = []
  private genes: Gene[] = []
  public fitness = 0

  constructor(private genome_history: GenomeHistory) {
    this.inputs = genome_history.inputs
    this.outputs = genome_history.outputs

    for (let i = 0; i < this.inputs; i++) {
      this.nodes.push(new Node(this.total_nodes, 0))
      this.total_nodes++
    }

    for (let i = 0; i < this.outputs; i++) {
      this.nodes.push(new Node(this.total_nodes, 1))
      this.total_nodes++
    }
  }

  public debug_info() {
    console.log("[Genome] Size Nodes: ", this.nodes.length)
    console.log("[Genome] Size Genes: ", this.genes.length)
  }

  public clone() {
    const clone = new Genome(this.genome_history)
    clone.total_nodes = this.total_nodes
    clone.fitness = this.fitness

    clone.nodes = this.nodes.map(node => node.clone())
    clone.genes = this.genes.map(gene => gene.clone())

    clone.connect_genes()
    return clone
  }

  private connect_nodes(n1: Node, n2: Node) {
    const n1_layer = n1.layer !== 1 ? n1.layer : 1000000
    const n2_layer = n2.layer !== 1 ? n2.layer : 1000000

    if (n1_layer > n2_layer) {
      [n1, n2] = [n2, n1]
    }

    const c = this.genome_history.exists(n1, n2)
    const x = new Gene(n1, n2)

    if (c) {
      x.innovation = c.innovation
      if (!this.exists(x.innovation)) {
        this.genes.push(x)
      }
    }
    else {
      x.innovation = this.genome_history.global_innovation
      this.genome_history.global_innovation++
      this.genome_history.all_genes.push(x)
      this.genes.push(x)
    }
  }

  private add_gene() {
    let n1 = this.nodes[Math.floor(Math.random() * this.nodes.length)]
    let n2 = this.nodes[Math.floor(Math.random() * this.nodes.length)]

    while (n1.layer === n2.layer || n2.layer === 0) {
      n1 = this.nodes[Math.floor(Math.random() * this.nodes.length)]
      n2 = this.nodes[Math.floor(Math.random() * this.nodes.length)]
    }

    this.connect_nodes(n1, n2)
  }

  private add_node() {
    if (this.genes.length == 0) this.add_gene()
    if (Math.random() < .2) this.genome_history.highest_hidden++

    const layer = Math.floor(Math.random() * (this.genome_history.highest_hidden - 1)) + 2
    const node = new Node(this.total_nodes, layer)
    this.total_nodes++

    let gene = this.genes[Math.floor(Math.random() * this.genes.length)]
    let l1 = gene.in_node.layer
    let l2 = gene.out_node.layer

    if (l2 == 1) l2 = 1000000

    while (l1 > node.layer || l2 < node.layer) {
      gene = this.genes[Math.floor(Math.random() * this.genes.length)]
      l1 = gene.in_node.layer
      l2 = gene.out_node.layer
      if (l2 == 1) l2 = 1000000
    }

    this.connect_nodes(gene.in_node, node)
    this.connect_nodes(node, gene.out_node)

    const last_gene = this.genes.at(-1)
    if (last_gene) last_gene.weight = 1.0

    const second_to_last = this.genes.at(-2)
    if (second_to_last) second_to_last.weight = gene.weight

    gene.enabled = false
    this.nodes.push(node)
  }

  public mutate() {
    if (this.genes.length == 0) this.add_gene()
    if (Math.random() < .8) this.genes.forEach(gene => gene.mutate())
    if (Math.random() < .08) this.add_gene()
    if (Math.random() < .02) this.add_node()
  }

  private get_node(node_number: number): Node | undefined {
    return Profiler.time("get_node", () => {

      const node = this.nodes.find(node => node.number === node_number)
      
      if (!node) {
        throw new Error(`Node not found: Something's wrong — ${node_number}`)
      }
      
      return node
    })
  }

  private get_gene(innovation: number) {
    return Profiler.time("get_gene", () => {

      const gene = this.genes.find(gene => gene.innovation === innovation)
      
      if (!gene) {
        throw new Error(`Gene not found: Something's wrong — ${innovation}`)
      }
      
      return gene.clone()
    })
  }

  private connect_genes() {
    this.genes.forEach(gene => {
      const node_in = this.get_node(gene.in_node.number)
      const node_out = this.get_node(gene.out_node.number)

      if (!node_in || !node_out) {
        throw new Error("Gene connection failed due to missing node")
      }

      gene.in_node = node_in
      gene.out_node = node_out
    });

    this.nodes.forEach(node => node.genes = []);
    this.genes.forEach(gene => gene.out_node.genes.push(gene));
  }

  public get_outputs(nn_inputs: number[]) {
    if (nn_inputs.length !== this.inputs) {
      throw new Error("Wrong number of inputs")
    }


    this.nodes.forEach(node => {
      node.output = 0
      node.genes = []
    })

    for (let i = 0; i < this.inputs; i++) {
      this.nodes[i].output = nn_inputs[i]
    }


    this.connect_genes()

    for (let layer = 2; layer < this.genome_history.highest_hidden + 1; layer++) {
      const nodes_in_layers = this.nodes.filter(node => node.layer === layer)
      nodes_in_layers.forEach(node => node.calculate())
    }

    const final_outputs: number[] = []
    for (let i = this.inputs; i < this.inputs + this.outputs; i++) {
      this.nodes[i].calculate()
      final_outputs.push(this.nodes[i].output)
    }

    return final_outputs
  }

  private exists(innovation: number) {
    return Profiler.time("exists", () => this.genes.some(gene => gene.innovation == innovation))
  }

  private get_weight(innovation: number) {
    return Profiler.time("get_weight", () => {
      const gene = this.genes.find(gene => gene.innovation === innovation)
      return gene ? gene.weight : -1
    })
  }

  private get_highest_innovation(genes: Gene[]): number {
    return genes.length > 0
      ? Math.max(...genes.map(g => g.innovation))
      : 0
  }

  public crossover(partner: Genome) {
    const child = new Genome(this.genome_history)
    child.nodes = []

    const more_fit = this.fitness > partner.fitness ? this : partner;
    const less_fit = this.fitness > partner.fitness ? partner : this;

    if (this.total_nodes > partner.total_nodes) {
      child.total_nodes = this.total_nodes;
      for (let i = 0; i < this.total_nodes; i++) {
        child.nodes.push(this.nodes[i].clone());
      }
    } else {
      child.total_nodes = partner.total_nodes;
      for (let i = 0; i < partner.total_nodes; i++) {
        child.nodes.push(partner.nodes[i].clone());
      }
    }

    const p1_highest_innovation = this.get_highest_innovation(this.genes)
    const p2_highest_innovation = this.get_highest_innovation(partner.genes)
    const highest_innovation = Math.max(p1_highest_innovation, p2_highest_innovation);

    for (let i = 0; i <= highest_innovation; i++) {
      const e1 = this.exists(i);
      const e2 = partner.exists(i);

      if (e1 || e2) {
        if (e1 && e2) {
          const gene = Math.random() < 0.75 ? more_fit.get_gene(i) : less_fit.get_gene(i);
          child.genes.push(gene);
        } else {
          if (e1) child.genes.push(this.get_gene(i));
          if (e2) child.genes.push(partner.get_gene(i));
        }
      }
    }
    child.connect_genes()
    return child
  }

  private calculate_compatibility(partner) {
    const p1_highest_innovation = this.get_highest_innovation(this.genes)
    const p2_highest_innovation = this.get_highest_innovation(partner.genes)

    const highest_innovation = this.fitness > partner.fitness ? p1_highest_innovation : p2_highest_innovation

    let matching = 0
    let disjoint = 0
    let excess = 0

    let total_weights = 0
    let average_weights = 0

    for (let i = 0; i < highest_innovation; i++) {
      const e1 = this.exists(i)
      const e2 = partner.exists(i)

      if (e1 || e2) {
        if (e1 && e2) {
          matching++
          total_weights += Math.abs(this.get_weight(i)) + Math.abs(partner.get_weight(i))
          continue
        }
        disjoint++
      }
    }

    average_weights = total_weights / (matching === 0 ? 1 : matching)

    for (let i = highest_innovation + 1; i <= Math.max(p1_highest_innovation, p2_highest_innovation); i++) {
      const e1 = this.exists(i)
      const e2 = partner.exists(i)
      if (e1 || e2) {
        excess++
      }
    }

    const N = highest_innovation < 20 ? 1 : highest_innovation

    const c1 = 1.0
    const c2 = 1.0
    const c3 = 0.4

    const E = c1 * excess / N
    const D = c2 * disjoint / N
    const W = c3 * average_weights
    const delta = E + D + W
    return delta
  }
}

File: python/libs/neat/genome.py
from __future__ import annotations
from webbrowser import get
from .genome_history import *
from itertools import count
from .gene import *
from .node import *
import random



class Genome:
    def __init__(self, gh: GenomeHistory) -> None:
        self.genome_history = gh

        self.inputs = gh.inputs
        self.outputs = gh.outputs

        self.input_layer = 0
        self.output_layer = 10

        self.total_nodes = 0
        self.node_indexer = None

        self.nodes: list[Node] = []
        self.genes: list[Gene] = []

        self.fitness = 0.0
        self.adjusted_fitness = 0.0

        for _ in range(self.inputs):
            self.nodes.append(Node(self.total_nodes, 0))
            self.total_nodes += 1

        for _ in range(self.outputs):
            self.nodes.append(Node(self.total_nodes, 1))
            self.total_nodes += 1

    def clone(self):
        clone = Genome(self.genome_history)

        clone.total_nodes = self.total_nodes
        clone.adjusted_fitness = self.adjusted_fitness
        clone.fitness = self.fitness

        clone.nodes.clear()
        clone.genes.clear()

        for i in range(len(self.nodes)):
            clone.nodes.append(self.nodes[i].clone())

        for i in range(len(self.genes)):
            clone.genes.append(self.genes[i].clone())

        clone.connect_genes()
        return clone

    def connect_nodes(self, n1: Node, n2: Node):
        n1_layer = n1.layer if n1.layer != 1 else 1000000
        n2_layer = n2.layer if n2.layer != 1 else 1000000

        if n1_layer > n2_layer:
            n1, n2 = n2, n1

        c = self.genome_history.exists(n1, n2)
        x = Gene(n1, n2)

        if c:
            x.innovation = c.innovation
            if not self.exists(x.innovation):
                self.genes.append(x)
        else:
            x.innovation = self.genome_history.global_innovation
            self.genome_history.global_innovation += 1
            self.genome_history.all_genes.append(x)
            self.genes.append(x)

    def mutate(self):
        if len(self.genes) == 0:
            self.add_gene()

        if random.random() < .5:  # add gene
            self.add_gene()
        if random.random() < 1:  # add node
            self.add_node()
        if random.random() < 0:  # remove gene
            self.remove_gene()
        if random.random() < 1:  # remove node
            self.remove_node()

        for i in range(len(self.genes)):
            self.genes[i].mutate()

    def add_gene(self):
        n1 = random.choice(self.nodes)
        n2 = random.choice(self.nodes)

        while n1.layer == n2.layer or (n2.layer == 0):
            n1 = random.choice(self.nodes)
            n2 = random.choice(self.nodes)

        self.connect_nodes(n1, n2)

    def add_node(self):
        if len(self.genes) == 0:
            self.add_gene()

        enabled_genes = [g for g in self.genes if g.enabled]
        print(enabled_genes)

        if not enabled_genes:
            print("no genes to split")
            return

        gene = random.choice(enabled_genes)
        get_node_id = self.get_new_node_key()
        node = Node(get_node_id, (gene.in_node.layer + gene.out_node.layer) / 2)

        self.total_nodes += 1
        self.connect_nodes(gene.in_node, node)
        self.connect_nodes(node, gene.out_node)

        self.genes[-1].weight = 1.0
        self.genes[-2].weight = gene.weight
        gene.enabled = False
        self.nodes.append(node)

    def remove_node(self):
        if (self.inputs + self.outputs >= len(self.nodes)):
            return

        nodes = self.nodes[self.inputs:-self.outputs]
        node_to_delete = random.choice(nodes)

        genes_to_delete: list[Gene] = []
        for gene in self.genes:
            if (gene.in_node.number == node_to_delete.number or gene.out_node.number == node_to_delete.number):
                genes_to_delete.append(gene)

        for gene in genes_to_delete:
            self.genes.remove(gene)

        self.nodes.remove(node_to_delete)
        self.total_nodes -= 1

        for i, node in enumerate(self.nodes):
            node.number = i

    def remove_gene(self):
        if (len(self.genes) == 0):
            return

        if self.genes:
            delete_index = random.randint(0, len(self.genes) - 1)
            self.genes.pop(delete_index)

    def get_node(self, n: int):
        for i in range(len(self.nodes)):
            if self.nodes[i].number == n:
                return self.nodes[i]

        raise Exception(f"Node not found, number: {n}")

    def get_gene(self, innovation: int) -> Gene:
        for g in self.genes:
            if g.innovation == innovation:
                return g.clone()

        raise Exception(f"Gene not found, innovation: {innovation}")

    def get_new_node_key(self):
        if self.node_indexer is None:
            self.node_indexer = count(len(self.nodes) + 1)

        new_id = next(self.node_indexer)

        return new_id

    def connect_genes(self):

        for i in range(len(self.genes)):
            self.genes[i].in_node = self.get_node(self.genes[i].in_node.number)
            self.genes[i].out_node = self.get_node(self.genes[i].out_node.number)

        for i in range(len(self.nodes)):
            self.nodes[i].genes.clear()

        for i in range(len(self.genes)):
            self.genes[i].out_node.genes.append(self.genes[i])

    def get_outputs(self, inputs: list[float]) -> list[float]:
        if len(inputs) != self.inputs:
            raise RuntimeError("Wrong number of inputs")

        for node in self.nodes:
            node.output = 0
            node.genes = []

        for i in range(self.inputs):
            self.nodes[i].output = inputs[i]

        self.connect_genes()

        for layer in range(2, self.genome_history.highest_hidden + 1):
            nodes_in_layer: list[Node] = []
            for n in range(len(self.nodes)):
                if self.nodes[n].layer == layer:
                    nodes_in_layer.append(self.nodes[n])

            for n in range(len(nodes_in_layer)):
                nodes_in_layer[n].calculate()

        final_outputs: list[float] = []
        for n in range(self.inputs, self.inputs + self.outputs):
            self.nodes[n].calculate()
            final_outputs.append(self.nodes[n].output)

        return final_outputs

    def exists(self, innovation: int):
        for c in self.genes:
            if c.innovation == innovation:
                return True
        return False

    def get_weight(self, innovation: int):
        for g in self.genes:
            if g.innovation == innovation:
                return g.weight
        return -1

    def crossover(self, partner: Genome) -> Genome:
        child = Genome(self.genome_history)
        child.nodes.clear()

        try:
            p1_highest_innovation = max([(x.innovation) for x in self.genes])
        except Exception:
            p1_highest_innovation = 0

        try:
            p2_highest_innovation = max([(x.innovation) for x in partner.genes])
        except Exception:
            p2_highest_innovation = 0

        if self.total_nodes > partner.total_nodes:
            child.total_nodes = self.total_nodes
            for node in self.nodes:
                child.nodes.append(node.clone())
        else:
            child.total_nodes = partner.total_nodes
            for node in partner.nodes:
                child.nodes.append(node.clone())

        highest_innovation = (
            p1_highest_innovation if self.fitness > partner.fitness else p2_highest_innovation
        )

        for i in range(highest_innovation + 1):
            e1 = self.exists(i)
            e2 = partner.exists(i)
            if e1 or e2:
                if e1 and e2:
                    gene = (
                        self.get_gene(i)
                        if (random.random()) < 0.5
                        else partner.get_gene(i)
                    )
                    child.genes.append(gene)
                    continue
                if e1:
                    child.genes.append(self.get_gene(i))
                if e2:
                    child.genes.append(partner.get_gene(i))

        child.connect_genes()
        return child

    def calculate_compatibility(self, partner: Genome):
        try:
            p1_highest_innovation = max([(x.innovation) for x in self.genes])
        except Exception:
            p1_highest_innovation = 0

        try:
            p2_highest_innovation = max([(x.innovation) for x in partner.genes])
        except Exception:
            p2_highest_innovation = 0

        if self.fitness > partner.fitness:
            highest_innovation = p1_highest_innovation
        else:
            highest_innovation = p2_highest_innovation

        matching = 0
        disjoint = 0
        excess = 0

        total_weights = 0
        average_weights = 0

        for i in range(highest_innovation + 1):
            e1 = self.exists(i)
            e2 = partner.exists(i)
            if e1 or e2:
                if e1 and e2:
                    matching += 1
                    total_weights += abs(self.get_weight(i)) + abs(partner.get_weight(i))
                    continue
                disjoint += 1

        average_weights = total_weights / (1 if matching == 0 else matching)

        for i in range(highest_innovation + 1, max(p1_highest_innovation, p2_highest_innovation) + 1):
            e1 = self.exists(i)
            e2 = partner.exists(i)
            if e1 or e2:
                excess += 1

        N = 1 if highest_innovation < 20 else highest_innovation

        c1 = 1.0
        c2 = 1.0
        c3 = 0.4

        E = c1 * excess / N
        D = c2 * disjoint / N
        W = c3 * average_weights

        delta = E + D + W

        return delta
